agent-control/src/sub_agent/on_host/supervisor.rs

```diff
@@ -214,244 +199,222 @@ impl NotStartedSupervisorOnHost {
+                }
 
-            restart_policy.backoff(|duration| {
-                // early exit if supervisor timeout is canceled
-                wait_exit_timeout(ctx.clone(), duration);
-            });
+                span.exit();
+            }
         };
 
         NotStartedThreadContext::new(executable_data.bin.clone(), callback).start()
     }
 }
 
-/// From the `ExitStatus`, send appropriate event and emit logs.
-fn handle_termination(
-    exec_data: &ExecutableData,
-    exit_status: ExitStatus,
-    health_publisher: EventPublisher<(String, HealthWithStartTime)>,
-    agent_id: &AgentID,
+/// Waits for the command to complete or be cancelled
+fn wait_exit(
+    mut command: CommandOSStarted,
     bin: &str,
-    start_time: SystemTime,
-) {
-    if !exit_status.success() {
-        debug!(%exit_status, "Informing of executable as unhealthy");
-        let last_error = format!(
-            "path '{}' with args '{}' failed with '{}'",
-            exec_data.bin,
-            exec_data.args.join(" "),
-            exit_status
-        );
-        let unhealthy: Unhealthy = Unhealthy::new(last_error).with_status(format!(
-            "process exited with code: {:?}",
-            exit_status.code().unwrap_or_default()
-        ));
-
-        if let Err(err) = health_publisher.publish((
-            exec_data.id.to_string(),
-            HealthWithStartTime::new(unhealthy.into(), start_time),
-        )) {
-            error!("Error publishing health status for {}: {err}", exec_data.id);
+    stop_consumer: &EventConsumer<CancellationMessage>,
+    healthy_publish_delay: Duration,
+    health_handler: &HealthHandler,
+) -> Result<(ExitStatus, bool), CommandError> {
+    info!("Waiting for executable to complete or be cancelled");
+    let mut was_cancelled = false;
+    let deadline = Instant::now() + healthy_publish_delay;
+    let mut already_published = false;
+
+    // Busy waiting is avoided with `is_cancelled_with_timeout`
+    while command.is_running() {
+        // Shutdown the spawned process when the cancel signal is received.
+        // This ensures the thread stops in time.
+        if stop_consumer.is_cancelled_with_timeout(WAIT_FOR_EXIT_TIMEOUT) {
+            info!(supervisor = bin, "Stopping executable");
+            if let Err(err) = command.shutdown() {
+                error!(supervisor = bin, "Failed to stop executable: {err}");
+            }
+            info!(supervisor = bin, msg = "Executable terminated");
+            was_cancelled = true;
         }
 
-        error!(
-            %agent_id,
-            supervisor = bin,
-            exit_code = ?exit_status.code(),
-            "supervisor process exited unsuccessfully"
-        )
+        // Publish healthy status once after the process has been running
+        // for an arbitrary long time without issues.
+        if !already_published && Instant::now() > deadline {
+            debug!("Informing executable as healthy");
```

Comment:
Nit: maybe adding some information regarding `healthy_publish_delay` could be useful in the debug message.

Reply:
done